keps/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,170 @@
+  - [Jobframwork](#jobframwork)
+  - [Batch/Job](#batchjob)
+- [Testing Plan](#testing-plan)
+  - [NonRegression](#nonregression)
+  - [Unit Tests](#unit-tests)
+  - [Integration tests](#integration-tests)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This proposal allows Kueue to reclaim the resources of a successfully completed Pod of a Workload even before the whole Workload completes its execution. Freeing the unused resources of a Workload earlier may allow more pending Workloads to be admitted. This will allow Kueue to utilize the existing resources of the cluster much more efficiently.
+
+In the remaining of this document:
+1. *Job* refers to any kind job supported by Kueue including `batch/job`, `MPIJob`, `RayJob`, etc.
+2. *Pods of a Workload* means the same as *Pods of a Job* and vice-versa.
+
+
+## Motivation
+
+Currently, the quota assigned to a Job is reclaimed by Kueue only when the whole Job finishes.
+For Jobs with multiple Pods, the resources will be reclaimed only after the last Pod of the Job finishes.
+This is not efficient as the Jobs might have different needs during execution, for instance the needs of a `batch/job` having the `parallelism` equal to `completions` will decrease with every Pod finishing it's execution.
+
+### Goals
+
+- Utilize the unused resources of the successfully completed Pods of a running Job.
+
+### Non-Goals
+
+- Preempting a Workload or Pods of a Workload to free resources for a pending Workload.
+- Partially admitting a Workload.
+- Monitor the Job's pod execution.
+
+## Proposal
+
+Reclaiming the resources of the succeeded Pods of a running Job as soon as the Pod completes its execution.
+
+We propose to add a new field `.status.ReclaimablePods` to the Workload API. `.status.ReclaimablePods` is a list that holds the count of Pods belonging to a PodSet whose resources are no longer needed and could be reclaimed by Kueue.
+
+### Pod Completion Accounting
+
+Since the ability of actively monitoring pods execution of a job is not in the scope of the core Kueue implementation, the number of pods for which the resources are no longer needed for each PodSet should be reported by each framework specific `GenericJob` implementation.
+
+For this purpose the `GenericJob` interface should be changed to either:
+
+1. Add an additional method able to report this
+
+```go
+type GenericJob interface {
+    // ...
+
+    // Get reclaimable pods.
+    ReclaimablePods() []ReclaimablePod
+}
+```
+
+2. Modify the current `Finished` method
+
+```go
+type GenericJob interface {
+    // ...
+
+    // Get reclaimable pods.
+    Finished() (condition metav1.Condition, reclaimablePods map[string]int)
+}
+```
+
+#### Reference design for (`batch/Job`)
+
+Having a job defined with **P** `parallelism` and **C** `completions`, and **n** number of completed pod executions,
+the expected result should be:
+
+```go
+map[string]int{
+    "main": P - min(P,(C-n))
+}
+```
+### API
```

Comment:
At a high level, reclaiming resources makes sense for the most of the time, but not always that case like some jobs may want to restart all the pods when running into a pod failure, reclaiming all the resources may lead to the job end in failure because insufficient resources . So maybe we can make it an option, reclaim or not, default to reclaim.

Reply:
For workload this should not be a problem, however this per-job opt-out needs to be done at job level, so the only way I can think of is  by using some "special" annotation/label.

Should we do that?

Reply:
We can add when needed, not conflict with current API.